### ISSUE TITLE
Rubinius and Yosemite: obsolete FAQ (and possibly unfriendly build system)

### ISSUE THREAD

**Issue by lloeki** [rubinius/rubinius#3286]
One must now install llvm 3.4 `brew install homebrew/versions/llvm34` and build with `ruby-install rbx -- --llvm-path="/usr/local/Cellar/llvm34/3.4.2"`, but the homebrew tap makes llvm-config & al. suffixed, and thus available as `llvm-config-3.4`. The configure step thus fails because the build system is unable to pick it up.

**Comment by brixen** [comment]
@lloeki which version of OS X are you using? Currently, `brew install llvm` should install 3.5 and although it is keg-only, Rubinius configure will find `llvm-config` via `brew list llvm`.

**Comment by lloeki** [comment]
Yosemite (see issue title). `brew install llvm` does install 3.5, but it turns out I was hellbent into installing 2.2.10 (being the last version reported by `ruby-install`), which required llvm 3.4, and I was landing on the FAQ page.

The "possibly unfriendly build system" was referring to precisely this case, as I suppose it should be able to pick keg-only versions from homebrew-versions tap too (someday llvm 3.6 will be released and install will be bound to fail again)

All of this is notwithstanding the fact that given all of this new info, currently installing rbx 2.5.0 quickly fails with:

```
~> rm -rf ~/src 
~> ruby-install rbx 2.5.0
>>> Installing rbx 2.5.0 into /Users/lloeki/.rubies/rbx-2.5.0 ...
>>> Installing dependencies for rbx 2.5.0 ...
Warning: openssl-1.0.1l already installed
Warning: readline-6.3.8 already installed
Warning: libyaml-0.1.6_1 already installed
Warning: gdbm-1.11 already installed
>>> Downloading http://releases.rubini.us/rubinius-2.5.0.tar.bz2 into /Users/lloeki/src ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 11.1M  100 11.1M    0     0  4103k      0  0:00:02  0:00:02 --:--:-- 4103k
>>> Verifying rubinius-2.5.0.tar.bz2 ...
*** No md5 checksum for /Users/lloeki/src/rubinius-2.5.0.tar.bz2
*** No sha1 checksum for /Users/lloeki/src/rubinius-2.5.0.tar.bz2
*** No sha256 checksum for /Users/lloeki/src/rubinius-2.5.0.tar.bz2
*** No sha512 checksum for /Users/lloeki/src/rubinius-2.5.0.tar.bz2
>>> Extracting rubinius-2.5.0.tar.bz2 to /Users/lloeki/src/rubinius-2.5.0 ...
>>> Bundling rubinius 2.5.0 ...
Fetching: bundler-1.7.12.gem (100%)
Successfully installed bundler-1.7.12
Parsing documentation for bundler-1.7.12
Installing ri documentation for bundler-1.7.12
Done installing documentation for bundler after 3 seconds
1 gem installed
Fetching gem metadata from https://rubygems.org/..
Using rake 10.4.2
Installing daedalus-core 0.2.0
Installing redcard 1.1.0
Installing rubinius-ast 2.3.1
Installing rubinius-bridge 1.1.0
Installing rubinius-compiler 2.3.0
Installing rubinius-melbourne 2.3.0.0
Installing rubinius-processor 2.3.0
Installing rubinius-toolset 2.3.1
Installing rubinius-build_tools 2.0.0
Using bundler 1.7.12
Updating files in vendor/cache
Your bundle is complete!
Use `bundle show [gemname]` to see where a bundled gem is installed.
>>> Configuring rubinius 2.5.0 ...
Checking gcc: found
Checking g++: found
Checking bison: found

Detected old configuration settings, forcing a clean build
  Checking for 'llvm-config': /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/spec_set.rb:92:in `block in materialize': Could not find rake-10.4.2 in any of the sources (Bundler::GemNotFound)
	from /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/spec_set.rb:85:in `map!'
	from /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/spec_set.rb:85:in `materialize'
	from /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/definition.rb:132:in `specs'
	from /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/definition.rb:177:in `specs_for'
	from /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/definition.rb:166:in `requested_specs'
	from /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/environment.rb:18:in `requested_specs'
	from /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/runtime.rb:13:in `setup'
	from /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler.rb:122:in `setup'
	from /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/setup.rb:17:in `<top (required)>'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
not found
ABORT: unable to set up LLVM

Running 'configure' failed. Please check configure.log for more details.
!!! Configuration of rbx 2.5.0 failed!
```

when it has obviously been bundled by `ruby-install` right before.

So, multiple issues stacking up here. I'll try to sort things out before complaining again ;-). Sorry for the noise. I'll reopen a more precise issue if need be.

**Comment by lloeki** [comment]
Both of the following worked:

    PATH="$(brew --prefix llvm)/bin:$PATH" ruby-install rbx 2.4.1
    PATH="$(brew --prefix llvm)/bin:$PATH" ruby-install rbx 2.5.0

Stumbled on [ruby-install#184](https://github.com/postmodern/ruby-install/issues/184) and #3247.

Also, seems like `export RUBYGEMS_GEMDEPS='-'` was wrecking some havoc, at least with 2.4.1. Well, whatever, I've got it installed in my peculiar environment.

**Comment by jemc** [comment]
@lloeki 
If you look at the last traceback you showed, you'll see it includes the fragment:
`Could not find rake-10.4.2 in any of the sources (Bundler::GemNotFound)`.

You probably got this error by not running `bundle install` with the "build ruby" to be sure you had the gems needed to build.  The hope is that soon we'll be building without Ruby as a dependency, and much of what makes the building-from-source process confusing or hard to approach will go away.

Meanwhile, installing with `ruby-install` is a great option for not having to worry about much of this.

**Comment by lloeki** [comment]
@jemc The trace is me using `ruby-install` (after unsetting `RUBYGEMS_GEMDEPS`, which makes bundler totally trip up). Neither `ruby-install rbx 2.4.1` nor `ruby-install rbx 2.5.0`  can find `llvm-config`, even though it is listed in `brew list llvm`, and forcing `PATH` makes it work. The `Could not find rake-10.4.2` is totally weird (esp.  since `Using rake 10.4.2` appears in the `bundle` phase) but appears only for 2.5.0, and only when `llvm-config` can't be found.

Unless someone else can reproduce this, I'll consider this a fluke on my environment.

**Comment by brixen** [comment]
@lloeki I really don't think you're system is failing to find LLVM. This is what I see in you output:

```
Checking for 'llvm-config': /Users/lloeki/.gem/ruby/2.2.0/gems/bundler-1.7.12/lib/bundler/spec_set.rb:92:in `block in materialize': Could not find rake-10.4.2 in any of the sources (Bundler::GemNotFound)
```

It's purely coincidental that configure is checking for LLVM at that point. If it were doing something else, that other thing would fail, but the real failure is this issue with the Rake version.

What Ruby switcher are you using?

**Comment by lloeki** [comment]
> What Ruby switcher are you using?

chruby (together with ruby-install), although the behaviour is the same without chruby.

> the real failure is this issue with the Rake version.

This failure reproducibly does not happen when `llvm-config` is found. In fact I just reran it, and this time it complained about the `daedalus-core-0.2.0` gem instead of rake.

Here is the output of a failing 2.5.0 `configure.log`:

```
[2015-01-22 17:42:49] Configuring Rubinius...
[2015-01-22 17:42:49] Command line: "--prefix=/Users/lloeki/.rubies/rbx-2.5.0 --with-opt-dir=/usr/local/opt/openssl:/usr/local/opt/readline:/usr/local/opt/libyaml:/usr/local/opt/gdbm"
[2015-01-22 17:42:49] Checking gcc:
[2015-01-22 17:42:49]  found
[2015-01-22 17:42:49] Checking g++:
[2015-01-22 17:42:49]  found
[2015-01-22 17:42:49] Checking bison:
[2015-01-22 17:42:49]  found
[2015-01-22 17:42:49] ---

Detected old configuration settings, forcing a clean build
---
[2015-01-22 17:42:49]   Checking for 'llvm-config': 
[2015-01-22 17:42:49] not found
[2015-01-22 17:42:49]*** ERROR ABORT: unable to set up LLVM
```

This is the same output as for rbx 2.4.1, except rbx 2.4.1 doesn't raise an error about a missing gem.

To sum up:

- both 2.4.1 and 2.5.0 fail in the same way in configure when they can't find llvm-config
- only 2.5.0 raises an error about a random gem, and only when it can't find llvm-config
- both 2.4.1 and 2.5.0 build and install correctly when provided the path to llvm-config, and none complain about a missing gem

**Comment by sgerrand** [comment]
@brixen: I've just encountered the same issue that @lloeki detailed above, with some extra quirks due to Homebrew's packaging.

OS: Mac OS X 10.10.3
Rubinius: 2.5.5
llvm: 3.5.1

As the Homebrew `llvm35` package appends `-3.5` to the LLVM binaries, I had to create a symlink from `llvm-config-3.5` to `llvm-config`. Once that was done, prepending the `ruby-install` invocation with `PATH="$(brew --prefix llvm35)/bin:$PATH"` enabled a successful installation of rbx-2.5.5.

**Comment by brixen** [comment]
@sgerrand there is a Homebrew binary for 2.5.5:

```
$ brew update
$ brew tap rubinius/apps
$ brew install rubinius
```

If you have a `~/.rubies` directory, the Homebrew recipe will copy 2.5.5 there as well so you can use it with chruby.

**Comment by sgerrand** [comment]
@brixen: Brilliant, thanks for letting me know!

**Comment by thbar** [comment]
@brixen thanks a lot for this! As some feedback, I tried to install rubinius (to verify compatibility of a gem I maintain) but went through multiple hurdles (using rvm, downloading directly etc). I only found your snippet after googling `ABORT: unable to set up LLVM` etc. Maybe adding the homebrew tip somewhere on the download page or in the getting started would be nice? (if it's there I missed it, sorry!).

**Comment by brixen** [comment]
@thbar yes, we need to have better instructions on getting binary builds of Rubinius. If you have time to put together a PR, the website is in https://github.com/rubinius/rubinius.github.io.
